File: days/day17.py
#!/usr/bin/env python
import re


def vm_run(prog, regs):
    regs = list(regs)
    ra, rb, rc = 4, 5, 6

    def _arg(v):
        match v:
            case 4 | 5 | 6:
                return regs[v - 4]
            case _:
                return v

    def _set(r, v):
        regs[r - 4] = v

    pn = len(prog)
    so = list()
    ip = 0
    while 0 <= ip < pn:
        op, val = prog[ip:ip+2]
        match op:
            case 0:
                _set(ra, _arg(ra) >> _arg(val))
                ip += 2
            case 1:
                _set(rb, _arg(rb) ^ val)
                ip += 2
            case 2:
                _set(rb, _arg(val) % 8)
                ip += 2
            case 3:
                if _arg(ra):
                    ip = val
                else:
                    ip += 2
            case 4:
                _set(rb, _arg(rb) ^ _arg(rc))
                ip += 2
            case 5:
                so.append(_arg(val) % 8)
                ip += 2
            case 6:
                _set(rb, _arg(ra) >> _arg(val))
                ip += 2
            case 7:
                _set(rc, _arg(ra) >> _arg(val))
                ip += 2
            case _:
                raise Exception(f'unhandled {op=!r} {arg=!r}')
    return so


def disasm(prog, file=None):
    comb = {4: 'ra', 5: 'rb', 6: 'rc'}
    for ip in range(0, len(prog), 2):
        op, val = prog[ip:ip+2]
        arg = comb.get(val, val)
        print(f'{ip:03o}: {op} {val}  ', end='', file=file)
        match op:
            case 0:
                print(f'(adv) ra <<= {arg}', file=file)
            case 1:
                print(f'(bxl) rb ^= {val}', file=file)
            case 2:
                print(f'(bst) rb = {arg} % 8', file=file)
            case 3:
                print(f'(jnz) {val}', file=file)
            case 4:
                print(f'(bxc) rb ^= rc', file=file)
            case 5:
                print(f'(out) {arg} % 8', file=file)
            case 6:
                print(f'(bdv) rb = ra << {arg}', file=file)
            case 7:
                print(f'(cdv) rc = ra << {arg}', file=file)
            case _:
                raise Exception(f'unhandled {op=!r} {val=!r}')
 
def vm_parse(data):
    regs = list(map(int, re.findall(r'[ABC]: (\d+)', data)))
    prog = list(map(int, re.findall(r'\d+', data.split('Program:')[-1])))
    return prog, regs


def part1(data):
    prog, regs = vm_parse(data)
    res = vm_run(prog, regs)
    ans = ','.join(map(str, res))
    return ans


def part2(data):
    def rev(prog, acc):
        if not prog:
            yield acc
        else:
            p = prog[-1]
            for i in range(8):
                a = ((acc << 3) | i) >> (i ^ 3)
                if (a % 8) ^ i == p:
                    yield from rev(prog[:-1], (acc << 3) | i)

    prog, regs = vm_parse(data)
    #disasm(prog)
    ans = None
    for ans in rev(prog, 0):
        regs[0] = ans
        res = vm_run(prog, regs)
        if res == prog:
            break
    return ans


data = '''
Register A: 729
Register B: 0
Register C: 0

Program: 0,1,5,4,3,0
'''
assert part1(data) == '4,6,3,5,6,3,5,2,1,0'


def handle_disasm(args):
    data = args.file.read()
    prog, _ = vm_parse(data)
    disasm(prog)


def handle_sat(args):
    import z3
    data = args.file.read()
    prog, _ = vm_parse(data)

    opt = z3.Optimize()
    s = z3.BitVec('s', len(prog) * 3)
    ra, rb, rc = s, 0, 0

    for x in prog:
        for ip in range(0, len(prog), 2):
            comb = {4: ra, 5: rb, 6: rc}
            op, val = prog[ip:ip+2]
            arg = comb.get(val, val)
            match op:
                case 0:
                    ra = ra / (1 << arg)
                case 1:
                    rb = rb ^ val
                case 2:
                    rb = arg % 8
                case 3:
                    pass
                case 4:
                    rb = rb ^ rc
                case 5:
                    opt.add((arg % 8) == x)
                case 6:
                    rb = ra / (1 << arg)
                case 7:
                    rc = ra / (1 << arg)
    opt.add(ra == 0)

    res = opt.check()
    assert res == z3.sat, repr(res)
    ans = opt.model().eval(s).as_long()

    assert vm_run(prog, [ans,0,0]) == prog
    print(ans)


def main(args):
    fp = args.file if 'file' in args else open('day17.in')
    data = fp.read()
    ans = part1(data)
    print('part1:', ans)
    ans = part2(data)
    print('part2:', ans)


if __name__ == '__main__':
    import argparse
    parser = argparse.ArgumentParser()
    parser.set_defaults(func=main)
    subp = parser.add_subparsers()

    filep = subp.add_parser('file')
    filep.add_argument('file', nargs='?', type=argparse.FileType(), default='day17.in', help='problem text')
    filep.set_defaults(func=main)

    dasmp = subp.add_parser('disasm', aliases=['dasm', 'asm'])
    dasmp.add_argument('file', nargs='?', type=argparse.FileType(), default='day17.in', help='problem text')
    dasmp.set_defaults(func=handle_disasm)

    sat = subp.add_parser('z3', aliases=['sat'])
    sat.add_argument('file', nargs='?', type=argparse.FileType(), default='day17.in', help='problem text')
    sat.set_defaults(func=handle_sat)

    args = parser.parse_args()
    args.func(args)

File: days/day17.jl
#!/usr/bin/env julia


function _disasm(img)
    a,b,c,prog... = img
    println("A=$(a) B=$(b) C=$(c)")
    println("prog=", prog)
    _sarg = "0123ABC"
    pc = 0
    while pc + 2 <= length(prog)
        op, larg = prog[pc+1:pc+2]
        sarg = _sarg[larg + 1]
        pc += 2
        if op == 0
            println("$(pc-2): adv $(sarg)")
        elseif op == 1
            println("$(pc-2): bxl $(larg)")
        elseif op == 2
            println("$(pc-2): bst $(sarg)")
        elseif op == 3
            println("$(pc-2): jnz $(larg)")
        elseif op == 4
            println("$(pc-2): bxc")
        elseif op == 5
            println("$(pc-2): out $(sarg)")
        elseif op == 6
            println("$(pc-2): bdv $(sarg)")
        elseif op == 7
            println("$(pc-2): cdv $(sarg)")
        else
            println("$(pc-2): $(op) $(arg)")
            throw(error("unhandled opcode $(op)"))
        end
    end
end


function _arg(regs, arg)
    if arg in 0:3
        arg
    elseif arg in 4:6
        regs[arg - 3]
    end
end


function _run(img)
    a,b,c,prog... = img
    pc = 0
    out = Int[]
    while pc + 2 <= length(prog)
        op, larg = prog[pc+1:pc+2]
        carg = _arg([a,b,c], larg)
        pc += 2
        if op == 0
            a ÷= 2^carg
        elseif op == 1
            b ⊻= larg
        elseif op == 2
            b = carg % 8
        elseif op == 3
            (a != 0) && (pc = larg)
        elseif op == 4
            b ⊻= c
        elseif op == 5
            push!(out, carg % 8)
        elseif op == 6
            b = a ÷ 2^carg
        elseif op == 7
            c = a ÷ 2^carg
        else
            throw(error("unhandled opcode $(op)"))
        end
    end
    img[1:3] = [a, b, c]
    return out
end


function part1(data)
    img = [parse(Int, m.match) for m in eachmatch(r"\d+", data)]
    out = _run(img)
    ans = join(out, ",")
    return ans
end


function part2(data)
    img = [parse(Int, m.match) for m in eachmatch(r"\d+", data)]
    prog = img[4:end]
    out = empty(prog)
    a = 1
    while true
        img[1] = a
        out = _run(img)
        out == prog && return a
        if out == prog[end-length(out)+1:end]
            a *= 8
        else
            a += 1
        end
    end
end


data = """
Register A: 729
Register B: 0
Register C: 0

Program: 0,1,5,4,3,0
"""
@assert part1(data) == "4,6,3,5,6,3,5,2,1,0"


data = """
Register A: 2024
Register B: 0
Register C: 0

Program: 0,3,5,4,3,0
"""
#@assert part2(data) == 117440


data = readchomp("day17.in")
println("part1: ", part1(data))
println("part2: ", part2(data))
